### ISSUE TITLE
There are no tests in this project, use_rezero=True is non-functional

### ISSUE THREAD

**Issue by fcampagne** [lucidrains/performer-pytorch#28]
Tests are needed to validate that models can train in various configurations. 
I built and ran simple tests (trying to get authorization to contribute as a PR) and found that use_rezero=True kills the gradient and results in a performer model that cannot learn. The fix consists in initializing the rezero parameter with a small value, but not zero (e.g., 1E-3 works in my tests). Zero prevents any signal to pass through the module so that the parameter will never change from zero.

**Comment by lucidrains** [comment]
I see, very interesting

Is this your own finding or something in the rezero paper that I missed?

**Comment by fcampagne** [comment]
This is from my own tests. I looked for a mention in the preprint, but did not see this mentioned. I guess that’s the secret sauce recipe people keep to themselves.

**Comment by fcampagne** [comment]
The test code is simple. You create a model, record the parameters, do one step of optimization (e.g., with SGD) with a large learning rate (0.1) and check which parameters have changed. If you enable rezero as implemented in your repo you will see that none of the parameters change after the update. I also checked that the loss stays approximately constant over training on a real problem, while the loss can decrease when use_rezero=False.

**Comment by lucidrains** [comment]
@fcampagne ok, i'll roll with this https://github.com/lucidrains/performer-pytorch/commit/aa726b5e0c42df7ec163c41632a798298738a472

there's very few people in the world looking at these experimental features. one other guy told me he's had good luck with Rezero by using different layer-wise learning rates. I've found rezero to work on full-attention (qk)v projects, but start failing with more diverse dataset corpuses

glad to hear you were able to make it work!

**Comment by lucidrains** [comment]
@fcampagne rezero is probably one of the most experimental features i keep around, just because certain users say they like it. i assure you the rest of the features are all ones that should work quite well!

**Comment by fcampagne** [comment]
And I guess with no test, nobody can tell easily if/when things break down. Would you be open to PRs that add tests if I get approval to submit?

**Comment by lucidrains** [comment]
Yeah sure

Well, this whole repo may need more testing.. we don't even know if Performer works for sure yet

**Comment by etienne87** [comment]
i do not understand ReZero:

```
class ReZero(nn.Module):
    def __init__(self, fn):
        super().__init__()
        self.g = nn.Parameter(torch.tensor(1e-3))
        self.fn = fn

    def forward(self, x, **kwargs):
        return self.fn(x, **kwargs) * self.g
```
From the paper, shouldn't this be:
```
return self.fn(x, **kwargs) * self.g + x
```

**Comment by etienne87** [comment]
also, how to combine ReZero + Reversible Layers?

**Comment by etienne87** [comment]
by the way thanks for all this pile of awesomeness in code
